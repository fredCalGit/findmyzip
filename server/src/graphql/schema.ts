import { gql } from "apollo-server";

export const typeDefs = gql`
  type Location {
    postCode: String
    country: String
    countryAbbreviation: String
    places: [Place]
  }
  type Place {
    placeName: String
    state: String
    stateAbbreviation: String
    longitude: String
    latitude: String
  }
  type Query {
    location(country: String, postCode: String): Location
  }
`;
