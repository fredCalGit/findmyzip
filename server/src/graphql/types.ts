export interface InputArgs {
  country: string;
  postCode: string;
}

export type Location = {
  postCode: string;
  country: string;
  countryAbbreviation: string;
  places: [Place];
};

export type Place = {
  placeName: string;
  state: string;
  stateAbbreviation: string;
  longitude: string;
  latitude: string;
};

export type LocationResult = {
  "post code": string;
  country: string;
  "country abbreviation": string;
  places: [PlaceResult];
};

export type PlaceResult = {
  "place name": string;
  state: string;
  "state abbreviation": string;
  longitude: string;
  latitude: string;
};
