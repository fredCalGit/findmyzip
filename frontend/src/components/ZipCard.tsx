import React, { useState } from "react";
import { gql, useQuery } from "@apollo/client";
import { Button, Card, CardActions, CardContent } from "@mui/material";
import { Form } from "./Form";
import { Header } from "./Header";
import { SearchTabs } from "./SearchTabs";

export const ZipCard = () => {
  const [queryVariables, setQueryVariables] = useState<QueryVariables>({
    country: "US",
    postCode: "0000000",
  });
  const [cachedVariables, setCachedVariables] = useState<
    QueryVariables[] | null
  >(null);

  const { data } = useQuery(query, {
    variables: queryVariables,
    errorPolicy: "all",
  });

  const handleSubmit = ({
    country,
    postCode,
  }: {
    country: string;
    postCode: string;
  }) => {
    setQueryVariables({ country, postCode });

    if (!cachedVariables) {
      setCachedVariables([{ country, postCode }]);
    } else {
      if (cachedVariables.length >= 5) {
        cachedVariables.splice(0, 1);
        setCachedVariables([...cachedVariables, { country, postCode }]);
      } else {
        setCachedVariables([...cachedVariables, { country, postCode }]);
      }
    }
  };

  return (
    <Card sx={{ height: "80vh" }}>
      <Header />
      <CardContent>
        <Form handleSubmit={handleSubmit} />
        <CardActions>
          <Button
            size="small"
            onClick={() => setCachedVariables(null)}
            disabled={!cachedVariables}
            color="error"
          >
            Clear History
          </Button>
        </CardActions>

        <SearchTabs
          data={data}
          cachedVariables={cachedVariables}
          setQueryVariables={setQueryVariables}
        />
      </CardContent>
    </Card>
  );
};

type QueryVariables = {
  country: string;
  postCode: string;
};

const query = gql`
  query GetLocations($country: String, $postCode: String) {
    location(country: $country, postCode: $postCode) {
      postCode
      country
      countryAbbreviation
      places {
        placeName
        state
        stateAbbreviation
        longitude
        latitude
      }
    }
  }
`;
