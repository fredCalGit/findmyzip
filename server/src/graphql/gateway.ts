import { RESTDataSource } from "apollo-datasource-rest";
import { InputArgs } from "./types";

interface Locations extends RESTDataSource {
  getLocation: ({ country, postCode }: InputArgs) => Promise<any>;
}

export class LocationsAPI extends RESTDataSource implements Locations {
  constructor() {
    super();
    this.baseURL = "http://api.zippopotam.us/";
  }

  async getLocation({ country, postCode }: InputArgs) {
    return this.get(`${country}/${postCode}`, { country, postCode });
  }
}
