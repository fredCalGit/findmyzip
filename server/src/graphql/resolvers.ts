export const resolvers = {
  Query: {
    location: async (_source, { country, postCode }, { dataSources }) => {
      try {
        let result = await dataSources.locationsAPI.getLocation({
          country: country.toUpperCase(),
          postCode,
        });
        let response = {
          ...result,
          postalCode: result["post code"],
          countryAbbreviation: result["country abbreviation"],
          places: result.places.map((place) => {
            return {
              ...place,
              placeName: place["place name"],
              stateAbbreviation: place["state abbreviation"],
            };
          }),
        };
        return response;
      } catch (err) {
        console.log(err);
        throw new Error("ZipCode not found!");
      }
    },
  },
};
