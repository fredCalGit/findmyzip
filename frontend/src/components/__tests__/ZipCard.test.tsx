import React from "react";
import { render } from "@testing-library/react";
import { ZipCard } from "../ZipCard";
import { ApolloProvider } from "@apollo/client";
import { client } from "../../graphql";

test("renders the ZipCard", () => {
  const { container } = render(
    <ApolloProvider client={client}>
      <ZipCard />
    </ApolloProvider>
  );
  expect(container).toMatchSnapshot();
});
