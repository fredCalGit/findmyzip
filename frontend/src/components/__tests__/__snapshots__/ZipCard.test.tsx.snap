// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the ZipCard 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-18x3soy-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-1qvr50w-MuiTypography-root"
        >
          findMyZip
        </span>
      </div>
      <div
        class="MuiCardHeader-action css-sgoict-MuiCardHeader-action"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-whjmaz-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="Brightness4Icon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69zM12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6c3.31 0 6 2.69 6 6s-2.69 6-6 6z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <div
        class="MuiBox-root css-1vt08d7"
      >
        <div
          class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon css-bs2jcm-MuiAutocomplete-root"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for="country-input"
              id="country-input-label"
            >
              Choose a Country
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-autocomplete="list"
                aria-expanded="false"
                aria-invalid="false"
                autocapitalize="none"
                autocomplete="off"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                id="country-input"
                role="combobox"
                spellcheck="false"
                type="text"
                value="United States"
              />
              <div
                class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
              >
                <button
                  aria-label="Clear"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-clearIndicator css-1glvl0p-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-clearIndicator"
                  tabindex="-1"
                  title="Clear"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                    data-testid="CloseIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
                <button
                  aria-label="Open"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                  tabindex="-1"
                  title="Open"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="ArrowDropDownIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-14lo706"
                >
                  <span>
                    Choose a Country
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root css-9ahdq0-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="zipcode-input"
            id="zipcode-input-label"
          >
            Enter Zip-Code
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="zipcode-input"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Enter Zip-Code
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Search
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textError MuiButton-sizeSmall MuiButton-textSizeSmall Mui-disabled MuiButton-root MuiButton-text MuiButton-textError MuiButton-sizeSmall MuiButton-textSizeSmall css-6ekzlo-MuiButtonBase-root-MuiButton-root"
          disabled=""
          tabindex="-1"
          type="button"
        >
          Clear History
        </button>
      </div>
      <div
        class="MuiBox-root css-8atqhb"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
        />
      </div>
    </div>
  </div>
</div>
`;
